### PR TITLE
Pupil Labs - Pupil LSL Relay Plugin

### PR DESCRIPTION
**Pull request by papr**
Pupil Capture plugin that relays data from the backbone to the lab streaming layer.

There are no additional dependencies (other than installing the LSL Python bindings) since all plugins have access to Pupil Capture's resources.

### CODE REVIEW

**Comment by dmedine** [comment]
This looks very good. Thanks for contributing!

I am hesitant to pull this into the repository, however, for a couple of reasons. The first is that there is already an app for streaming PupilPro data. I'm sure it is out of date and that yours is a vast improvement. However, some care should be taken not to have two apps for the same device. 

The second is that PupilLabs is probably going to roll out a version of their software that includes its own LSL outlets. I am of the opinion that it is always better to have vendors do this -- not that they're necessarily better at it (although they might be), but it means less work for the community here and it gets vendors more invested in LSL. Again, I want to avoid multiple solutions to the same problem.

Finally, there are some design choices that I would like to have a discussion about (if anyone is interested). It seems like you're app makes a lot of outlets. I think I understand why you did it this way. It is a simple way to manage whether or not to stream from both eyes or just one, etc. but it does make for a lot of streams sitting on the network. It's kind of an aesthetic judgement, but I prefer not to have labrecorder cluttered up with outlets that I may or may not need because I will inevitably check the wrong one and get bad data.

**Comment by papr** [comment]
TL;DR:
- This PR represents @pupil-labs official "native" LSL integration
- LSL outlets are not flexible enough to map Pupil Capture's data directly

---

Dear @dmedine 
My name is Pablo and I work as software developer for @pupil-labs (as previously mentioned in #148). Let me elaborate on how Capture works, and the difficulties with LSL that arise from that.

Headset

Above, you can see that a Pupil headset can have up to three cameras, the front camera (`world`) and two eye cameras (`eye.0` and `eye.1`). Pupil Capture processes each video stream in a separate process. The eye processes detect the pupil positions and report these to the `world` process. The `world` process is also responsible for running all plugins. Every feature, from the gaze mapper (that maps raw pupil positions to gaze positions on the `world` image) to simple visualisation features, are implemented as plugins.

Since version `v0.8`, Pupil Capture uses a [ZMQ](http://zeromq.org)-based message-passing architecture. It does not only allow us to write highly modular plugins but is also the recommended way to interact with our software as third-party developer. See our [wiki](https://github.com/pupil-labs/pupil/wiki/Pupil-Interprocess-and-Network-Communication) for more information. The PupilPro app does not support the zmq backend at all.

The data structures that we are using are not only hierarchical but are also not static, e.g gaze positions include their matching pupil positions (these can include data from `eye.0`, `eye.1`, or both) and (if activated) additional information from the 3D pupil detector. This makes a direct mapping of our data to LSL outlets very difficult. This is also the reason why there will not be a "native" LSL integration from us. LSL is just not flexible enough.

We know that LSL is a great tool that is used by a lot of researchers. That is why we tried to come up with a flexible way that not only allows researchers to have access to all our data (hence the Python representation outlets) but is also maintainable for us in the long run. This plugin is the closest thing to a "native" implementation that we can provide. It also allows the user to turn on and off outlets for pupil, gaze and notification data.
